app/models/spree/order.rb

```diff
@@ -615,6 +616,20 @@ def sorted_line_items
       end
     end
 
+    def can_generate_new_invoice?
+      return true if invoices.empty?
+
+      invoices.last.data["order"].to_json != to_json(include: :line_items)
```

Comment:
This check needs changing now, right? Or is `to_json` finding the right serializer already?

Reply:
I extracted the comparison logic into a separated class.
app/services/order_invoice_comparator.rb

I use a non-recursive BFS to compare invoices.